auth/realm/base/src/main/java/org/wildfly/security/auth/realm/JaasSecurityRealm.java

```diff
@@ -60,33 +63,60 @@
+    private static final String DEFAULT_CONFIGURATION_POLICY_TYPE = "JavaLoginConfig";
+    private final URI jaasConfigFilePath;
+    private final String entry;
     private final CallbackHandler handler;
+    private final ClassLoader classLoader;
 
     /**
      * Construct a new instance.
      *
-     * @param loginConfiguration the login configuration name to use
+     * @param jaasConfigFilePath path to JAAS configuration file
+     * @param classLoader        classLoader to use with LoginContext, this class loader must contain LoginModule CallbackHandler classes
      */
-    public JaasSecurityRealm(final String loginConfiguration) {
-        this(loginConfiguration, null);
+    public JaasSecurityRealm(final String jaasConfigFilePath, ClassLoader classLoader) {
+        this(jaasConfigFilePath, null, classLoader, null);
     }
 
     /**
      * Construct a new instance.
      *
-     * @param loginConfiguration the login configuration name to use
-     * @param handler the JAAS callback handler to use
+     * @param jaasConfigFilePath path to JAAS configuration file
+     * @param entry              login configuration file entry
+     * @param classLoader        classLoader to use with LoginContext, this class loader must contain LoginModule CallbackHandler classes
      */
-    public JaasSecurityRealm(final String loginConfiguration, final CallbackHandler handler) {
-        this.loginConfiguration = loginConfiguration;
-        this.handler = handler;
+    public JaasSecurityRealm(final String jaasConfigFilePath, final String entry, ClassLoader classLoader) {
+        this(jaasConfigFilePath, entry, classLoader, null);
+    }
+
+    /**
+     * Construct a new instance.
+     *
+     * @param jaasConfigFilePath path to JAAS configuration file
+     * @param entry              login configuration file entry
+     * @param callbackHandler    callbackHandler to pass to LoginContext
+     * @param classLoader        classLoader to use with LoginContext, this class loader must contain LoginModule CallbackHandler classes
+     */
+    public JaasSecurityRealm(final String jaasConfigFilePath, final String entry, final ClassLoader classLoader, final CallbackHandler callbackHandler) {
+        Assert.checkNotNullParam("entry", entry);
+        if (jaasConfigFilePath != null) {
+            this.jaasConfigFilePath = Paths.get(jaasConfigFilePath).toUri();
+        } else {
+            this.jaasConfigFilePath = null;
+        }
+        this.entry = entry;
+        this.handler = callbackHandler;
+        if (classLoader != null) {
+            this.classLoader = classLoader;
+        } else {
+            this.classLoader = Thread.currentThread().getContextClassLoader();
+        }
     }
 
     @Override
     public RealmIdentity getRealmIdentity(final Principal principal) {
-        return principal instanceof NamePrincipal ? new JaasRealmIdentity(principal) : RealmIdentity.NON_EXISTENT;
+        return new JaasRealmIdentity(principal);
```

Comment:
@darranl I do not map any specific subject's principal to be the main "caller" principal. The main caller principal is the one that was passed into the realm for authentication.